scipy/stats/_distn_infrastructure.py

```diff
@@ -3855,3 +3873,437 @@ def get_distribution_names(namespace_pairs, rv_base_class):
+    shape_info = dist._get_shape_info()
+
+    n_shapes = len(shape_info)
+    if shape_bounds is None or np.array(shape_bounds).size == 0:
+        shape_bounds = {}
+
+    if isinstance(shape_bounds, dict):
+        shape_bounds = shape_bounds.copy()  # don't mutate the user's object
+        shape_bounds_array = np.empty((n_shapes, 2))
+        for i in range(n_shapes):
+            shape_name = shape_info[i].shape_name
+            shape_bound = (shape_bounds.pop(shape_name, None)
+                           or shape_info[i].domain)
+            shape_bounds_array[i] = shape_bound
+        if shape_bounds:
+            message = ("Bounds provided for the following unrecognized shapes "
+                       f"will be ignored: {', '.join(shape_bounds.keys())}")
+            warnings.warn(message, RuntimeWarning, stacklevel=2)
+        shape_bounds = shape_bounds_array
+    else:
+        shape_bounds = np.asarray(shape_bounds, dtype=float)
+        if shape_bounds.shape != (n_shapes, 2):
+            message = (f"`shape_bounds` must have {n_shapes} elements, tuples "
+                       "containing the lower and upper bounds for the "
+                       f"following shape parameters: {dist.shapes}")
+            raise ValueError(message)
+
+    validated_shape_bounds = []
+    for i in range(n_shapes):
+        name = shape_info[i].shape_name
+        user_bounds = shape_bounds[i]
+        shape_domain = shape_info[i].domain
+        integral = shape_info[i].integrality
+        bound_type = "shape_bounds"
+        combined = _combine_bounds(name, bound_type, user_bounds, shape_domain,
+                                   integral)
+        validated_shape_bounds.append(combined)
+
+    # loc_bounds and scale_bounds input validation
+    loc_bounds = _combine_bounds("loc", "loc_bounds", loc_bounds,
+                                 (-np.inf, np.inf), discrete)
+    loc_bounds = [loc_bounds]
+
+    if discrete and scale_bounds is not None:
+        message = (f"`{dist.name}` is an instance of `rv_discrete`, which "
+                   "does not have `scale`. `scale_bounds` will be ignored.")
+        warnings.warn(message, RuntimeWarning, stacklevel=2)
+    if discrete:
+        scale_bounds = []
+    else:
+        scale_bounds = [_combine_bounds("scale", "scale_bounds", scale_bounds,
+                                        (0, np.inf), False)]
+
+    bounds = validated_shape_bounds + loc_bounds + scale_bounds
+    integrality = ([shape.integrality for shape in shape_info]
+                   + loc_scale_integrality)
+
+    # --- MLE Fitting --- #
+    def nllf(free_params, data=data):  # bind data NOW
+        with np.errstate(invalid='ignore', divide='ignore'):
+            return dist._penalized_nnlf(free_params, data)
+
+    with np.errstate(invalid='ignore', divide='ignore'):
+        if np.any(integrality):
+            res = optimizer(nllf, bounds, integrality=integrality)
```

Comment:
I guess single shot optimization works ok with univariate distributions. I was thinking that we could also do a multi-stage optimization if it had trouble to optimize a large search space. We know that parameters do not have the same sensitivity on the objective so we could separate in 2 or more groups of importance the parameters to optimize. It also really depends on the optimizer (like if gradients are involved and we don't have the same scale on the parameters)

Reply:
The default optimizer for now is `differential_evolution`, but `optimizer` gives the user full control over this sort of thing. Maybe [this comment](https://github.com/scipy/scipy/issues/11948#issuecomment-1008246023) is along the lines of what you had in mind for multi-stage optimization? (See the `zoom_opt` function.)